### PR TITLE
Add v9, v8plus, and leoncasa target feature to sparc and use v8plus in create_object_file

### LINKED ISSUE #132585: Our 32-bit Sparc handling isn't quite correct yet
See:
- https://github.com/rust-lang/rust/issues/130172
- https://github.com/rust-lang/rust/pull/131222
- https://github.com/rust-lang/rust/pull/132552

We need to, for compiling for 32-bit SPARC, distinguish when we should use the appropriate Elf Machine for V8 vs. V8+. This is necessary because this is the primary way of detecting instruction incompatibility for the architecture: providing a CPUID via instruction or register read at runtime hadn't become fashionable yet.

### PR DESCRIPTION
**Pull request by taiki-e**
This adds the following three unstable target features:

- `v9`: SPARC-V9 instructions ([LLVM definition][sparc-v9])
  - Relevant to https://github.com/rust-lang/rust/pull/131222#issuecomment-2453310963
  - Relevant to https://github.com/rust-lang/rust/pull/132472#discussion_r1832606081
  - This is also needed to implement https://github.com/taiki-e/atomic-maybe-uninit/pull/31 (depends on inline assembly support) more robustly.
- `v8plus`: SPARC-V8+ ABI ([LLVM definition][sparc-v8plus])
  - This is added in LLVM 20. In LLVM 19 and older, it is emulated to work the same way as LLVM in each LLVM version.
  - See https://github.com/rust-lang/rust/issues/132585#issuecomment-2453926257 for more.
- `leoncasa`: CASA instruction[^1] of LEON3 and LEON4 processors ([LLVM definition][sparc-leoncasa], LLVM feature name: `hasleoncasa`)
  - This is needed to implement https://github.com/taiki-e/atomic-maybe-uninit/pull/31 (depends on inline assembly support) more robustly.

cc @thejpster @glaubitz

r? workingjubilee

Closes #132585

@rustbot label +O-SPARC +A-target-feature

[^1]: Atomic CAS instruction

[sparc-v9]: https://github.com/llvm/llvm-project/blob/f5e4ffaa49254706ad6fa209de8aec28e20f0041/llvm/lib/Target/Sparc/Sparc.td#L37-L39
[sparc-v8plus]: https://github.com/llvm/llvm-project/blob/f5e4ffaa49254706ad6fa209de8aec28e20f0041/llvm/lib/Target/Sparc/Sparc.td#L37-L39
[sparc-leoncasa]: https://github.com/llvm/llvm-project/blob/llvmorg-19.1.0/llvm/lib/Target/Sparc/LeonFeatures.td#L32-L37

### CODE REVIEW

File: compiler/rustc_target/src/target_features.rs [hunk 1]
```diff
@@ -502,6 +502,13 @@ const IBMZ_ALLOWED_FEATURES: &[(&str, Stability, ImpliedFeatures)] = &[
     // tidy-alphabetical-end
 ];
 
+const SPARC_ALLOWED_FEATURES: &[(&str, Stability, ImpliedFeatures)] = &[
+    // tidy-alphabetical-start
+    ("hasleoncasa", Unstable(sym::sparc_target_feature), &[]),
```
Comment (workingjubilee) [inline compiler/rustc_target/src/target_features.rs:507]: uhhh, can we call it `leoncasa`?

Comment (taiki-e) [reply compiler/rustc_target/src/target_features.rs]: Removing the `has` prefix makes sense to me. Renamed.

**Comment by glaubitz** [comment]
CC @psumbera @rorth @koachan @jrtc27 @andreas-gaisler

**Comment by taiki-e** [comment]
> `v8plus` is a purely ABI flag, intended to signal that the G and O registers are 64-bit. It is separated in LLVM because we want to be GCC-compatible and so need to cater to a particular usecase in which V9 instructions are desired, but all registers are to be kept 32-bit (effectively banning instructions like CASXA which needs 64-bit registers to operate). If Rust doesn't want to do this, it could always set (or delete) `v9` and `v8plus` together...

Thanks for the clarification. I think separating `v8plus` from `v9` is the correct behavior (eventually) for Rust as well, as we do not want the ABI to be changed by `-C target-cpu` (`-mtune`).

However, we are using LLVM 19 which ties `v9` and `v8plus` together and need a way to fix a bug in rustc that incorrectly always uses EM_SPARC32PLUS (https://github.com/rust-lang/rust/issues/130172).

In https://github.com/rust-lang/rust/pull/131222 we are trying to adopt the same approach as in LLVM 19, but actually adding the `v8plus` feature and mapping them to the `v9` feature in LLVM 19 may be a better way here. (The implementation of it is in https://github.com/rust-lang/rust/commit/6f9d2a316ed8b79e00c6d9be931261f9e4f5a2c9 and even if we merge the LLVM 19's approach, we will probably need it when updating to LLVM 20.)

**Comment by taiki-e** [comment]
Added commit to add `v8plus` target feature and use it (the same one that mentioned in https://github.com/rust-lang/rust/pull/132552#issuecomment-2453463726). This fixes https://github.com/rust-lang/rust/issues/132585 and should work the same way as LLVM in each LLVM version. (See https://github.com/rust-lang/rust/issues/132585#issuecomment-2453926257 for more info.)

(If it would be better to separate that commit into a separate PR, I would be happy to separate it.)

**Comment by workingjubilee** [comment]
I think fixing that here should be fine?

Can we get a separate tracking issue for the SPARC target features though?

**Comment by taiki-e** [comment]
> Can we get a separate tracking issue for the SPARC target features though?

Opened https://github.com/rust-lang/rust/issues/132783 and updated the tracking issue number.

**Comment by workingjubilee** [comment]
Cool, thanks!

I reviewed this PR and it seems correct.

But also this PR should only affect tier 3 targets, so... yolo!

File: compiler/rustc_codegen_llvm/src/llvm_util.rs [hunk 2]
```diff
@@ -619,6 +628,8 @@ pub(crate) fn global_llvm_features(
             .features
             .split(',')
             .filter(|v| !v.is_empty() && backend_feature_name(sess, v).is_some())
+            // Drop +v8plus feature introduced in LLVM 20.
+            .filter(|v| *v != "+v8plus" || get_version() >= (20, 0, 0))
```
Comment (RalfJung) [inline compiler/rustc_codegen_llvm/src/llvm_util.rs:632]: This line here looks wrong... such cases should be handled in `to_llvm_feature`. And you did add that case there as well. So why did you also add it here? What makes this feature different from all the others that do not get a special case here?

Comment (RalfJung) [reply compiler/rustc_codegen_llvm/src/llvm_util.rs:632]: Ah, these are the LLVM feature names from the target, they don't go through `to_llvm_feature`, I see.

**Pull request merged**